drivers/can/mcp2515.c

```diff
@@ -0,0 +1,639 @@
+				       ~canintf);
+
+		/* check that no new interrupts happened while clearing known
+		 * ones */
+		mcp2515_cmd_read_reg(dev, MCP2515_ADDR_CANINTF, &canintf, 1);
+	}
+}
+
+static void mcp2515_int_thread(struct device *dev)
+{
+	struct mcp2515_data *dev_data = DEV_DATA(dev);
+
+	while (1) {
+		k_sem_take(&dev_data->int_sem, K_FOREVER);
+		mcp2515_handle_interrupts(dev);
+	}
+}
+
+static void mcp2515_int_gpio_callback(struct device *dev,
+				      struct gpio_callback *cb, u32_t pins)
+{
+	struct mcp2515_data *dev_data =
+		CONTAINER_OF(cb, struct mcp2515_data, int_gpio_cb);
+
+	k_sem_give(&dev_data->int_sem);
+}
+
+static const struct can_driver_api can_api_funcs = {
+	.configure = mcp2515_configure,
+	.send = mcp2515_send,
+	.attach_msgq = mcp2515_attach_msgq,
+	.attach_isr = mcp2515_attach_isr,
+	.detach = mcp2515_detach
+};
+
+
+static int mcp2515_init(struct device *dev)
+{
+	const struct mcp2515_config *dev_cfg = DEV_CFG(dev);
+	struct mcp2515_data *dev_data = DEV_DATA(dev);
+
+	k_sem_init(&dev_data->int_sem, 0, UINT_MAX);
+	k_mutex_init(&dev_data->tx_mutex);
+	k_sem_init(&dev_data->tx_sem, 3, 3);
+	k_sem_init(&dev_data->tx_cb[0].sem, 0, 1);
+	k_sem_init(&dev_data->tx_cb[1].sem, 0, 1);
+	k_sem_init(&dev_data->tx_cb[2].sem, 0, 1);
+	k_mutex_init(&dev_data->filter_mutex);
+
+	/* SPI config */
+	dev_data->spi_cfg.operation = SPI_WORD_SET(8);
+	dev_data->spi_cfg.frequency = dev_cfg->spi_freq;
+	dev_data->spi_cfg.slave = dev_cfg->spi_slave;
+
+	dev_data->spi = device_get_binding(dev_cfg->spi_port);
+	if (!dev_data->spi) {
+		LOG_ERR("SPI master port %s not found", dev_cfg->spi_port);
+		return -EINVAL;
+	}
+
+#ifdef CONFIG_CAN_MCP2515_GPIO_SPI_CS
+	dev_data->spi_cs_ctrl.gpio_dev =
+		device_get_binding(dev_cfg->spi_cs_port);
+	if (!dev_data->spi_cs_ctrl.gpio_dev) {
+		LOG_ERR("Unable to get GPIO SPI CS device");
+		return -ENODEV;
+	}
+
+	dev_data->spi_cs_ctrl.gpio_pin = dev_cfg->spi_cs_pin;
+	dev_data->spi_cs_ctrl.delay = 0U;
+
+	dev_data->spi_cfg.cs = &dev_data->spi_cs_ctrl;
+#else
+	dev_data->spi_cfg.cs = NULL;
+#endif  /* CAN_MCP2515_GPIO_SPI_CS */
+
+	/* Reset MCP2515 */
+	if (mcp2515_cmd_soft_reset(dev)) {
+		LOG_ERR("Soft-reset failed");
+		return -EIO;
+	}
+
+	/* Initialize interrupt handling  */
+	dev_data->int_gpio = device_get_binding(dev_cfg->int_port);
+	if (dev_data->int_gpio == NULL) {
+		LOG_ERR("GPIO port %s not found", dev_cfg->int_port);
+		return -EINVAL;
+	}
+
+	if (gpio_pin_configure(dev_data->int_gpio, dev_cfg->int_pin,
+			       (GPIO_DIR_IN | GPIO_INT | GPIO_INT_EDGE
+				| GPIO_INT_ACTIVE_LOW | GPIO_INT_DEBOUNCE))) {
+		LOG_ERR("Unable to configure GPIO pin %u", dev_cfg->int_pin);
+		return -EINVAL;
+	}
+
+	gpio_init_callback(&(dev_data->int_gpio_cb), mcp2515_int_gpio_callback,
+			   BIT(dev_cfg->int_pin));
+
+	if (gpio_add_callback(dev_data->int_gpio, &(dev_data->int_gpio_cb))) {
+		return -EINVAL;
+	}
+
+	if (gpio_pin_enable_callback(dev_data->int_gpio, dev_cfg->int_pin)) {
+		return -EINVAL;
+	}
+
```

Comment:
Can you add a bitrate option to Kconfig and add call mcp2515_configure here? The device should be initialized and ready to use after init.

Reply:
Done